cli/module_graph.rs

```diff
@@ -0,0 +1,725 @@
+use crate::tsc::get_available_libs;
+use deno_core::ErrBox;
+use deno_core::ModuleSpecifier;
+use futures::stream::FuturesUnordered;
+use futures::stream::StreamExt;
+use futures::Future;
+use futures::FutureExt;
+use serde::Serialize;
+use serde::Serializer;
+use std::collections::HashMap;
+use std::hash::BuildHasher;
+use std::pin::Pin;
+
+fn serialize_module_specifier<S>(
+  spec: &ModuleSpecifier,
+  s: S,
+) -> Result<S::Ok, S::Error>
+where
+  S: Serializer,
+{
+  s.serialize_str(&spec.to_string())
+}
+
+fn serialize_option_module_specifier<S>(
+  maybe_spec: &Option<ModuleSpecifier>,
+  s: S,
+) -> Result<S::Ok, S::Error>
+where
+  S: Serializer,
+{
+  if let Some(spec) = maybe_spec {
+    serialize_module_specifier(spec, s)
+  } else {
+    s.serialize_none()
+  }
+}
+
+#[derive(Debug, Serialize)]
+pub struct ModuleGraph(HashMap<String, ModuleGraphFile>);
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ImportDescriptor {
+  specifier: String,
+  #[serde(serialize_with = "serialize_module_specifier")]
+  resolved_specifier: ModuleSpecifier,
+  // These two fields are for support of @deno-types directive
+  // directly prepending import statement
+  type_directive: Option<String>,
+  #[serde(serialize_with = "serialize_option_module_specifier")]
+  resolved_type_directive: Option<ModuleSpecifier>,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ReferenceDescriptor {
+  specifier: String,
+  #[serde(serialize_with = "serialize_module_specifier")]
+  resolved_specifier: ModuleSpecifier,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ModuleGraphFile {
+  pub specifier: String,
+  pub url: String,
+  pub filename: String,
+  pub imports: Vec<ImportDescriptor>,
+  pub referenced_files: Vec<ReferenceDescriptor>,
+  pub lib_directives: Vec<ReferenceDescriptor>,
+  pub types_directives: Vec<ReferenceDescriptor>,
+  pub type_headers: Vec<ReferenceDescriptor>,
+  pub media_type: i32,
+  pub source_code: String,
+}
+
+type SourceFileFuture =
+  Pin<Box<dyn Future<Output = Result<SourceFile, ErrBox>>>>;
+
+pub struct ModuleGraphLoader {
+  permissions: Permissions,
+  file_fetcher: SourceFileFetcher,
+  maybe_import_map: Option<ImportMap>,
+  pending_downloads: FuturesUnordered<SourceFileFuture>,
+  pub graph: ModuleGraph,
+  is_dyn_import: bool,
+  analyze_dynamic_imports: bool,
+}
+
+impl ModuleGraphLoader {
+  pub fn new(
+    file_fetcher: SourceFileFetcher,
+    maybe_import_map: Option<ImportMap>,
+    permissions: Permissions,
+    is_dyn_import: bool,
+    analyze_dynamic_imports: bool,
+  ) -> Self {
+    Self {
+      file_fetcher,
+      permissions,
+      maybe_import_map,
+      pending_downloads: FuturesUnordered::new(),
+      graph: ModuleGraph(HashMap::new()),
+      is_dyn_import,
+      analyze_dynamic_imports,
+    }
+  }
+
+  pub async fn add_to_graph(
+    &mut self,
+    specifier: &ModuleSpecifier,
+  ) -> Result<(), ErrBox> {
```

Comment:
It's not immediately clear to me when this promise completes. Is it after the module is downloaded? Is it stored in the deno dir cache? It would be helpful to have some triple slash comment here.